### PR TITLE
Build json typegen files

### PR DESCRIPTION
**Pull request by AloeareV**
Build script is currently in a rudimentary proof-of-concept stage. We aren't currently using json_typegen at all, yet.

### CODE REVIEW

File: zcashrpc-macros/src/lib.rs [hunk 1]
```diff
@@ -1,6 +1,14 @@
...
 
 #[proc_macro]
-declare_rpc_response_structs(arg: TokenStream) -> TokenStream {
-    arg
+pub fn declare_rpc_client_methods(input: TokenStream) -> TokenStream {
+    calls::declare_rpc_client_methods(input.into()).into()
+}
+
+#[proc_macro]
+pub fn declare_rpc_response_types(_input: TokenStream) -> TokenStream {
```
Comment (zancas) [inline zcashrpc-macros/src/lib.rs:12]: Why do we have to stuff in `_input`?

File: build.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
+#[path = "src/client/utils.rs"]
+mod utils;
+
+//Copied from envelope.rs to not get more types and deps than we need
```
Comment (zancas) [inline build.rs:4]: Per your idea @AloeareV let's redefine modules such that we don't have copied code.

File: build.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let mut id = 0..;
+    let client = reqwest::Client::new();
+
+    //temporary POC, this will likely be replaced by a config-file in the
+    //short-term, and hopefully eventually a query against zcashd itself
+    let rpc_response_names: Vec<(&'static str, Vec<serde_json::Value>)> = vec![
```
Comment (zancas) [inline build.rs:28]: Eventually we'll automate the population of this list.

File: build.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        ("GetInfoResponse", Vec::new()),
+        ("GetBlockChainInfoResponse", Vec::new()),
+        ("GenerateResponse", vec![2.into()]),
+        ("ZGetNewAddressResponse", Vec::new()),
+    ];
+
+    //This assignment prevents the compiler from complaining if the file
+    //already exists. If the file already exists, this is correctly a noop,
+    //and not an error we need to be informed of.
```
Comment (zancas) [inline build.rs:37]: Eventually we'll need finer grained control.

File: build.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    let _json_dir = std::fs::DirBuilder::new().create("json_data");
+    for response_name in rpc_response_names {
+        //Logic mostly copied from client.rs, as we don't want to inherit deps
```
Comment (zancas) [inline build.rs:40]: We need a better way.

Comment (AloeareV) [reply build.rs]: We have a better way!

File: build.rs (hunk 2 above)
Comment (AloeareV) [reply build.rs]: We may want to move some things around again for clarity purposes, but I've done the movement needed to dry out the code!

File: zcashrpc-macros/src/lib.rs (hunk 1 above)
Comment (AloeareV) [reply zcashrpc-macros/src/lib.rs:12]: To match the interface expected by the `#[proc_macro]` attribute

File: build.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+mod utils;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    //temporary POC, this will likely be replaced by a config-file in the
+    //short-term, and hopefully eventually a query against zcashd itself
+    let rpc_response_names: Vec<(&'static str, Vec<serde_json::Value>)> = vec![
+        ("GetInfoResponse", Vec::new()),
+        ("GetBlockChainInfoResponse", Vec::new()),
+        ("GenerateResponse", vec![2.into()]),
```
Comment (zancas) [inline build.rs:11]: Can we change this to a `F::from`?

File: src/client/utils.rs [hunk 7]
```diff
@@ -33,3 +33,68 @@ pub fn get_zcashd_port() -> String {
...
+    ) -> RequestEnvelope {
+        RequestEnvelope {
+            id: id,
+            method: method,
+            params: params,
+        }
+    }
+}
+
+pub(crate) struct InnerCli {
```
Comment (zancas) [inline src/client/utils.rs:66]: I like this DRYer version.  I like the use of `pub(crate)` this seems like sound practice to me.

File: build.rs (hunk 6 above)
Comment (AloeareV) [reply build.rs:11]: Sure

File: build.rs (hunk 4 above)
Comment (AloeareV) [reply build.rs:17]: Issued!

File: build.rs (hunk 6 above)
Comment (AloeareV) [reply build.rs:11]: Issued!

**Comment by dannasessha** [comment]
#13 here builds upon #11. Testing this branch locally in anticipation of merge.

**Comment by dannasessha** [comment]
`cargo test` fails locally while running `zcashd` with `zcashd --regtest --printtoconsole`. 

Compiler errors include `unresolved imports`, `cannot find type` and `type annotation needed`.

**Comment by dannasessha** [comment]
`cargo test` passes on commit `0aa561a`

**Comment by dannasessha** [comment]
`cargo test` passes on commit `663ff0`: about to merge!

Just to make a note: at prior commit `0aa561`, `cargo test` would pass, but modified `Cargo.lock` file.

**Pull request merged**